### ISSUE TITLE
In Solr 7, take advantage use the new solr.data.home to separate cores from config

### ISSUE THREAD

**Issue by dsmiley** [docker-solr/docker-solr#133]
Solr 7 finally allows a convenient way to specify a data directory that does not have configuration in it.  https://issues.apache.org/jira/browse/SOLR-6671   Solr home, by default, has both. I think the Docker image should configure Solr in this way by default and then declare the data dir as a VOLUME.  We explored making solr.solr.home a VOLUME in the past but opted not to because the configs of a non-trivial size would get auto-copied to a data volume that would often go unused (if my memory serves).

**Comment by makuk66** [comment]
Heh; I spent some time yesterday experimenting with SOLR-6671, in the context of #130,  before reading your message here :-)

I had hoped this could make the need for the mycores directory and the INIT_SOLR_HOME handling go away. But I was sorely disappointed to learn that SOLR-6671 only deals with the data, not the core config, which still ends up in solr home. So, we still can't just mount a blank volume somewhere to create cores in. Sigh.

The use-case where `solr.data.home` might be nice is if you have the conf dir already, and want to mount it read-only, but keep data in a separate volume. E.g, with Solr 7 you can do (without any docker-solr changes):

```
# create a core by hand:
rm -fr mycore
mkdir mycore
docker run \
  -v "$PWD/mycore:/mycore" \
  --user "$(id -u):$(id -g)" \
  solr bash -c 'cp -r /opt/solr/server/solr/configsets/data_driven_schema_configs/* /mycore/'
touch mycore/core.properties

# create a directory for data
sudo rm -fr mydata
mkdir mydata
sudo chown 8983:8983 mydata # ignore this for the moment; that is fixable

# run docker
docker run \
  -p 8983:8983 \
  -v "$PWD/mycore:/opt/solr/server/solr/mycore:ro" \
  -v "$PWD/mydata:/opt/solr/server/solr/mydata" \
  -e SOLR_DATA_HOME=/opt/solr/server/solr/mydata \
  solr-7-rc2
```

but, it's not really a massive step up from what you can do with Solr 6, just mounting the data directory in the right location:

```
# create a core by hand:
rm -fr myconf
mkdir myconf
docker run \
  -v "$PWD/myconf:/myconf" \
  --user "$(id -u):$(id -g)" \
  solr:6 bash -c 'cp -r /opt/solr/server/solr/configsets/data_driven_schema_configs/conf/* /myconf/'

# create a directory for the core
sudo rm -fr mycore
mkdir mycore
touch mycore/core.properties
sudo chown -R 8983:8983 mycore # ignore this for the moment; that is fixable

# run docker
docker run \
  -p 8983:8983 \
  -v "$PWD/mycore:/opt/solr/server/solr/mycore" \
  -v "$PWD/myconf:/opt/solr/server/solr/mycore/conf:ro" \
  solr:6
```

or even just use a single directory and appropriate permissions:

```
# create a core by hand:
sudo rm -fr mycore
mkdir mycore
docker run \
  -v "$PWD/mycore:/mycore" \
  --user "$(id -u):$(id -g)" \
  solr:6 bash -c 'cp -r /opt/solr/server/solr/configsets/data_driven_schema_configs/* /mycore/'
chmod -R ugo-w mycore/conf
mkdir mycore/data
touch mycore/core.properties
chmod ugo-w mycore/core.properties mycore
sudo chown -R 8983:8983 mycore # ignore this for the moment; that is fixable

# run docker
docker run \
  -p 8983:8983 \
  -v "$PWD/mycore:/opt/solr/server/solr/mycore" \
  solr:6
```

You could argue that the best way of using docker-solr is in fact to follow the examples above: always mount explicit config and data dirs. Don't worry about solr-create/solr-precreate. But users have made it clear that those features are appreciated in some use-cases.

But I digress.


You're proposing defaulting `solr.data.home`, let's say for arguments sake to /opt/solr/server/solr/data, which would have to be created and chowned in the Dockerfile, and setting VOLUME for it.
That means existing users that currently mount full cores into /opt/solr/server/solr/somewhere will find that their data ends up not there, but in some random volume. Same for users using INIT_SOLR_HOME.

We could retain backwards compatibility with an alternative approach: leave the Dockerfile as-is, and then in the docker_entrypoint.sh we could do something like:

```
if [ ! -z "${NO_SOLR_DATA_HOME:-}" ]; then
  if [ -z "${SOLR_DATA_HOME:-}" ]; then
    DATA_HOME=/opt/solr/server/solr/data
    if [ -d "$DATA_HOME" ]; then
      export SOLR_DATA_HOME=$DATA_HOME
      fi
    fi
  fi
fi
``` 

But it's a bit hacky and non-standard (in the Solr world).
With this auto-detecting there still could be some subtle interplay with users who try to set SOLR_DATA_HOME in a custom solr.in.sh, so we might want to try and address that, but it will be fragile.

You mention VOLUME in the Dockerfile, and concerns about copying. I dislike VOLUME because it has inconsistent behaviour: if you volume mount on that location, copying happens, but if you bind mount on that location you don't. For empty directories the copying inconsistency is not a problem. But if you mount nothing, your data ends up in some volume where you must get at it with --volumes-from which is inconvenient. If you do mount, then VOLUME doesn't add anything.
I could just about accept it if we could designate empty data and logs volume, and have cores creates there, and no writing happens outside those directories, but as mentioned above, because of the core conf writing, that is not the case.

**Comment by dsmiley** [comment]
> But I was sorely disappointed to learn that SOLR-6671 only deals with the data, not the core config, which still ends up in solr home. So, we still can't just mount a blank volume somewhere to create cores in. Sigh.

Argh!  Man.  Forget solr.data.home then.

Reading through comments on SOLR-6671 I observed that Alan referred to a `coreRootDirectory` setting in solr.xml.  I think this is our ticket!  The only thing that would be nice is a way to auto-set it from a system property, and in turn set this from an environment variable.  Reading `SolrXmlConfig.fillSolrSection(...)` it'd be nice if it was generic to fall back on a system property if the value wasn't set for all these things, for that matter.

**Comment by makuk66** [comment]
> coreRootDirectory setting in solr.xml

To just try it out I did:

```
diff --git a/6.6/scripts/run-initdb b/6.6/scripts/run-initdb
index c64f1b9..d1d6481 100755
--- a/6.6/scripts/run-initdb
+++ b/6.6/scripts/run-initdb
@@ -20,6 +20,7 @@ set -e
 
 # init script for handling a custom SOLR_HOME
 /opt/docker-solr/scripts/init-solr-home
+/opt/docker-solr/scripts/set-core-root-directory
 
 # execute files in /docker-entrypoint-initdb.d before starting solr
 while read f; do
diff --git a/6.6/scripts/set-core-root-directory b/6.6/scripts/set-core-root-directory
new file mode 100755
index 0000000..e62becf
--- /dev/null
+++ b/6.6/scripts/set-core-root-directory
@@ -0,0 +1,18 @@
+#!/bin/bash
+#
+# Set the coreRootDirectory in solr.xml to SOLR_CORE_ROOT or mycores
+
+set -e
+
+if [[ "$VERBOSE" = "yes" ]]; then
+    set -x
+fi
+
+if [[ ! -z "$NO_SOLR_CORE_ROOT" ]]; then
+  exit 0
+fi
+
+SOLR_XML=/opt/solr/server/solr/solr.xml
+if [ -w "$SOLR_XML" ]; then
+  sed -i -e '/<solr>/ a <str name="coreRootDirectory">${SOLR_CORE_ROOT:/opt/solr/server/solr/mycores}</str>' "$SOLR_XML"
+fi
```

which modifies the solr.xml at run-time, but then `docker run -it -e VERBOSE=yes
  docker-solr/docker-solr:6.6 solr-create -c books` got me:

```
ERROR: Error CREATEing SolrCore 'books': Unable to create core [books] Caused by: Can't find resource 'solrconfig.xml' in classpath or '/opt/solr/server/solr/mycores/books'
```

Trying outside docker, with Solr 7, a manual solr.xml edit, I see the same thing.
There must be more to it.

**Comment by dsmiley** [comment]
I'll investigate and file a bug with Solr.

**Comment by makuk66** [comment]
That's about all we can do at the moment then I think. When new options arise, let's open a new issue

**Comment by morissm** [comment]
There's a lot of version specific stuff in default configsets and solr.xml and I'd rather not have them mixed in with my data if I can avoid it. It sometimes leads to some installations having slightly out of date configurations with weird bugs.

I'm perfectly fine with using solr.data.home and relying on zookeeper for configsets. In my case, the biggest problem is that core.properties files are not stored in solr.data.home which means that redeployments break core discovery.

I've filed a separate bug to that effect: https://issues.apache.org/jira/browse/SOLR-11508

**Comment by morissm** [comment]
@dsmiley I had read https://issues.apache.org/jira/browse/SOLR-9575 and had assumed that this is the bug you had promised to file.

I now see I got my chronology mixed up. Did you ever file that core.properties bug? If so, is there a patch incoming.

**Comment by dsmiley** [comment]
Sorry, I've been too busy to investigate further.  FWIW this is filed in my TODO list.
